File: typescript/src/safeFetch.ts
import { lookup } from "node:dns/promises";

/**
 * Hard limits the hardened implementation must enforce.
 * They are already wired into the tests — do not rename them.
 */
export const MAX_BYTES = 1 * 1024 * 1024; // 1 MiB
export const MAX_REDIRECTS = 2;
export const TIMEOUT_MS = 5000;

/** Thrown when a request is rejected for SSRF / policy reasons. */
export class SsrfError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "SsrfError";
  }
}

/**
 * TODO (student): return true for any IP an attacker could use to reach
 * internal infrastructure. You MUST cover, at minimum:
 *
 *   IPv4:
 *     - 127.0.0.0/8      (loopback)
 *     - 10.0.0.0/8       (private)
 *     - 172.16.0.0/12    (private)
 *     - 192.168.0.0/16   (private)
 *     - 169.254.0.0/16   (link-local, INCLUDING 169.254.169.254 cloud metadata!)
 *
 *   IPv6:
 *     - ::1              (loopback)
 *     - fc00::/7         (unique local)
 *     - fe80::/10        (link-local)
 *
 * The naive version below blocks nothing.
 */
export function isBlockedIp(_ip: string): boolean {
  // TODO: implement the checks above.
  return false;
}

/**
 * NAIVE, VULNERABLE implementation. It happily fetches whatever URL it is
 * given — attacker-controlled hosts, private IPs, redirects to the metadata
 * service, unbounded response bodies. Harden it until the security specs pass.
 *
 * Requirements to make the tests green:
 *   1. Only allow the `https:` protocol (reject http:, file:, gopher:, ...).
 *   2. Resolve the host with `lookup()` and reject when `isBlockedIp(address)`.
 *   3. Follow at most MAX_REDIRECTS redirects, RE-VALIDATING the host of every
 *      hop (an allowed host can 302 you to http://169.254.169.254).
 *   4. Enforce MAX_BYTES on the response body and TIMEOUT_MS on the request.
 *
 * Throw `SsrfError` (or any Error) when a request must be blocked.
 */
export async function safeFetch(rawUrl: string): Promise<Buffer> {
  // TODO: parse rawUrl and enforce protocol === "https:".
  // TODO: resolve the host and reject blocked IPs.
  // TODO: fetch with manual redirect handling (revalidate each hop).
  // TODO: enforce size + timeout while reading the body.

  const res = await fetch(rawUrl);
  const buf = Buffer.from(await res.arrayBuffer());
  return buf;
}

File: lib/safe_fetch.rb
# frozen_string_literal: true

require "net/http"
require "uri"
require "resolv"
require "ipaddr"

# SafeFetch downloads the body at a URL.
#
# WARNING: as shipped, this is the NAIVE / VULNERABLE version. It follows any
# scheme, any host (including internal ones), unlimited redirects, and imposes
# no size or time limits. It is the starting point for the DARE Labs lab
# "Secure a Download Endpoint Against SSRF".
#
# Complete the TODOs until spec/safe_fetch_spec.rb is green.
class SafeFetch
  # Raised when a request is refused for security reasons.
  class BlockedError < StandardError; end

  MAX_REDIRECTS = 2
  MAX_BYTES = 5 * 1024 * 1024 # 5 MiB
  TIMEOUT_SECONDS = 5

  def self.call(url)
    new.call(url)
  end

  def call(url, redirects_left = MAX_REDIRECTS)
    uri = URI.parse(url)

    # TODO (a): validate the scheme — only "https" should be allowed.
    #   Raise BlockedError for http:// and any other scheme.

    # TODO (b): resolve the host to an IP (Resolv.getaddress) and refuse to
    #   connect when it falls in a blocked range. Use private_ip? below.
    #   Remember 169.254.169.254 (cloud metadata) MUST be blocked.

    # NAIVE request: no timeout, no size cap. Replace with a guarded request.
    response = Net::HTTP.get_response(uri) # TODO (d): apply TIMEOUT_SECONDS.

    case response
    when Net::HTTPRedirection
      # TODO (c): follow at most MAX_REDIRECTS redirects, and RE-VALIDATE the
      #   new host (scheme + IP) on every hop. The naive version below follows
      #   blindly and forever.
      location = response["location"]
      return call(location, redirects_left - 1)
    else
      body = response.body.to_s

      # TODO (d): enforce MAX_BYTES — reject bodies larger than the limit
      #   (ideally while streaming, not after buffering everything).

      body
    end
  end

  private

  # Returns true when +ip+ (a String) is loopback, private, or link-local and
  # therefore must NOT be fetched.
  #
  # Blocked ranges:
  #   IPv4: 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16, 127.0.0.0/8,
  #         169.254.0.0/16 (includes 169.254.169.254 metadata)
  #   IPv6: ::1/128, fc00::/7, fe80::/10
  def private_ip?(ip)
    # TODO (b): implement using IPAddr, e.g.
    #   BLOCKED_RANGES.any? { |range| range.include?(IPAddr.new(ip)) }
    false
  end
end
